### PR TITLE
[Feature] Lightning integration example

### PR DESCRIPTION
**Pull request by svnv-svsv-jm**
## Description

This PR offers a convenient `lightning.pytorch.LightningModule` base class, from which one can inherit to be able to train a `torchrl` model using `lightning`.

## Motivation and Context

This PR is inspired by this issue: https://github.com/Lightning-Universe/lightning-bolts/issues/986

## Types of changes

What types of changes does your code introduce? Remove all that do not apply:

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds core functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (update in the documentation)
- [ ] Example (update in the folder of examples)

## Checklist

Go over all the following points, and put an `x` in all the boxes that apply.
If you are unsure about any of these, don't hesitate to ask. We are here to help!

- [x] I have read the [CONTRIBUTION](https://github.com/pytorch/rl/blob/main/CONTRIBUTING.md) guide (**required**)
- [x] My change requires a change to the documentation.
- [x] I have updated the tests accordingly (*required for a bug fix or a new feature*).
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Review by vmoens — DISMISSED** [review]
Pretty impressive work! It'll take me some time to review it.
Cc'ing people who may be interested 
@giadefa @albertbou92 @BY571 @tchaton

File: test/test_lightning_integration.py
Comment (vmoens) [inline test/test_lightning_integration.py:1]: I think this should belong to test_libs.py or test_trainer.py

File: test/test_lightning_integration.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    rollout = model.env.rollout(3)
+    print(f"Rollout of three steps: {rollout}")
+    print(f"Shape of the rollout TensorDict: {rollout.batch_size}")
+    print(f"Env reset: {model.env.reset()}")
+    print(f"Running policy: {model.policy_module(model.env.reset())}")
+    print(f"Running value: {model.value_module(model.env.reset())}")
+    # Collector
+    model.setup()
+    collector = model.train_dataloader()
+    for _, tensordict_data in enumerate(collector):
```
Comment (vmoens) [inline test/test_lightning_integration.py:33]: How long is this training loop?

File: test/test_lightning_integration.py [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+        n_mlp_layers=4,
+        use_checkpoint_callback=True,
+    )
+    # Rollout
+    rollout = model.env.rollout(3)
+    print(f"Rollout of three steps: {rollout}")
+    print(f"Shape of the rollout TensorDict: {rollout.batch_size}")
+    print(f"Env reset: {model.env.reset()}")
+    print(f"Running policy: {model.policy_module(model.env.reset())}")
+    print(f"Running value: {model.value_module(model.env.reset())}")
```
Comment (vmoens) [inline test/test_lightning_integration.py:29]: do we need these prints?

File: torchrl/lightning/loops.py [hunk 3]
```diff
@@ -0,0 +1,493 @@
...
+                Defaults to False.
+
+            num_envs (int, optional):
+                Number of environments to use.
+                If `num_envs > 1`, `num_envs` copies of the input env will be created.
+                Defaults to 1.
+
+            env_kwargs (ty.Dict[str, ty.Any], optional):
+                Parameters for your environment.
+                Defaults to {}.
```
Comment (vmoens) [inline torchrl/lightning/loops.py:126]:
```suggestion
            loss_module (TensorDictModule): A torchl :class:`~torchrl.objectives.LossModule` module.
            policy_module (TensorDictModule): Policy module.
            value_module (TensorDictModule): Value module.
            target_net_updater (SoftUpdate, optional): Target network updater. Defaults to ``None``.
            lr (float, optional): Learning rate. Defaults to ``3e-4``.
            max_grad_norm (float, optional): Maximum norm value for the gradients. Defaults to ``1.0``.
            frame_skip (int, optional): Frame skip value. Defaults to 1.
            frames_per_batch (int, optional): Number of frames per batch. Defaults to ``100``.
            total_frames (int, optional): Total number of frames. Defaults to ``100_000``.
            sub_batch_size (int, optional): Sub batch size. Defaults to ``1``.
            lr_monitor (str, optional): Name given to the loss to monitor.
                Defaults to ``"loss/train"``.
            lr_monitor_strict (bool, optional):
                Whether to raise an error in case the metric `lr_monitor` is not found.
                Defaults to ``False``.
            rollout_max_steps (int, optional): Max number of rollout steps.
                Defaults to ``1000``.
            automatic_optimization (bool, optional): Whether to use manual or automatic optimization.
                Defaults to ``True``.
            use_checkpoint_callback (bool, optional): Whether to use a checkpoint callback.
                Defaults to ``False``.
            save_every_n_train_steps (int, optional): Save every n steps. Defaults to ``100``.
            raise_error_on_nan (bool, optional): Whether to immediately raise an error if a NaN is encountered.
                Defaults to ``False``.
            num_envs (int, optional): Number of environments to use.
                If `num_envs > 1`, `num_envs` copies of the input env will be created.
                Defaults to ``1``.
            env_kwargs (ty.Dict[str, ty.Any], optional): Parameters for your environment.
                Defaults to ``{}``.
```

File: torchrl/lightning/loops.py [hunk 4]
```diff
@@ -0,0 +1,493 @@
...
+                save_last=True,
+                save_on_train_epoch_end=True,
+                every_n_train_steps=self.save_every_n_train_steps,
+            )
+            callbacks.append(ckpt_cb)
+        return callbacks
+
+    def configure_optimizers(self) -> OptimizerLRSchedulerConfig:
+        """Configures the optimizer (`torch.optim.Adam`) and the learning rate
+        scheduler (`torch.optim.lr_scheduler.CosineAnnealingLR`)."""
```
Comment (vmoens) [inline torchrl/lightning/loops.py:222]:
```suggestion
        """Configures the optimizer and the learning rate scheduler."""
```

File: torchrl/lightning/loops.py [hunk 5]
```diff
@@ -0,0 +1,493 @@
...
+
+    def on_train_batch_end(
+        self,
+        outputs: Tensor | ty.Mapping[str, ty.Any] | None,
+        batch: ty.Any,
+        batch_idx: int,
+    ) -> None:
+        """Check if we have to stop.
+        For some reason, Lightning can't understand this.
+        Probably because we are using an `IterableDataset`."""
```
Comment (vmoens) [inline torchrl/lightning/loops.py:283]: I don't get this docstring. Should it be a docstring or a comment?

File: torchrl/lightning/loops.py [hunk 6]
```diff
@@ -0,0 +1,493 @@
...
+        if clip_val is None:
+            clip_val = self.max_grad_norm
+        torch.nn.utils.clip_grad_norm_(self.loss_module.parameters(), clip_val)
+
+    def call_scheduler(self) -> None:
+        """Call schedulers. We are using an infinite datalaoder,
+        this will never be called by the `pl.Trainer` in the `on_train_epoch_end` hook.
+        We have to call it manually in the `training_step`."""
+        scheduler = self.lr_schedulers()
+        assert isinstance(scheduler, LRScheduler)
```
Comment (vmoens) [inline torchrl/lightning/loops.py:344]: no assert in the codebase

File: torchrl/lightning/loops.py [hunk 7]
```diff
@@ -0,0 +1,493 @@
...
+        ```
+        """
+
+    def step(
+        self,
+        batch: TensorDict,
+        batch_idx: int = 0,
+        tag: str = "train",
+    ) -> Tensor:
+        """Common step."""
```
Comment (vmoens) [inline torchrl/lightning/loops.py:371]: Can we expand this a bit?

File: torchrl/lightning/loops.py [hunk 8]
```diff
@@ -0,0 +1,493 @@
...
+            if "loss_" in key:
+                # If not finite, may raise error
+                if self.raise_error_on_nan:
+                    if value.isnan().any() or value.isinf().any():
+                        raise RuntimeError(f"Invalid loss value for {key}: {value}.")
+                # Update total loss
+                loss = loss + value
+                loss_dict[f"{key}/{tag}"] = value
+        # Sanity check and return
+        assert isinstance(loss, torch.Tensor)
```
Comment (vmoens) [inline torchrl/lightning/loops.py:428]: no assert in codebase

File: torchrl/lightning/ppo.py [hunk 9]
```diff
@@ -0,0 +1,123 @@
+"""Template for a PPO model on the pendulum env, for Lightning."""
+
+__all__ = ["PPOPendulum"]
```
Comment (vmoens) [inline torchrl/lightning/ppo.py:3]: missing header, no `__all__`

File: torchrl/lightning/ppo.py [hunk 10]
```diff
@@ -0,0 +1,123 @@
...
+
+import typing as ty
+
+import torch
+from tensordict import TensorDict  # type: ignore
```
Comment (vmoens) [inline torchrl/lightning/ppo.py:8]: why the 
```
# type: ignore
```
?

**Review by vmoens — COMMENTED** [review]
I'm leaning towards accepting a lightning backend for the trainers but given the limited bandwidth on my time I won't have much time to scale the PPO trainer to others. 

A recipe for other trainers would probably help others code their own trainer. For instance a tutorial (in a second PR)?
Are you considering adding other trainers?

Since this is a new "trainer" I think it should be moved to torchrl/trainers and we will need to make the 2 APIs somewhat compatible.

An example in `torchrl/examples` would be welcome!

I didn't do a very "in-depth" review, just pieces that I could easily spot for a more homogeneous formatting within the lib!

File: torchrl/lightning/ppo.py (hunk 10 above)
Comment (svnv-svsv-jm) [reply torchrl/lightning/ppo.py:8]: My `mypy` in my VSCode was complaining... So I mypy-ignored the line. This is not necessary in the codebase.

File: torchrl/lightning/_base.py [hunk 11]
```diff
@@ -0,0 +1,205 @@
+"""Creates a helper class for more complex models."""
+
+__all__ = ["BaseRL"]
+
+import typing as ty
+from loguru import logger
```
[the remark this answers is not in the document]
Comment (svnv-svsv-jm) [reply torchrl/lightning/_base.py:6]: Yep, forgot to remove `loguru`, which is used in the project I copied most of the code from. I will check out `torchrl`'s logger or remove logging entirely for this.

**Comment by svnv-svsv-jm** [comment]
I will take care of the comments, and add an example under `torchrl/examples`, but I will leave moving the code to `torchrl/trainers` for later.

File: torchrl/lightning/collector.py [hunk 12]
```diff
@@ -0,0 +1,83 @@
+# pylint: disable=abstract-method
```
[the remark this answers is not in the document]
Comment (svnv-svsv-jm) [reply torchrl/lightning/collector.py:1]: I get `Method '__getitem__' is abstract in class 'Dataset' but is not overridden in child class 'CollectorDataset'Pylint[W0223:abstract-method](https://pylint.readthedocs.io/en/latest/user_guide/messages/warning/abstract-method.html)`

File: torchrl/lightning/loops.py [hunk 13]
```diff
@@ -0,0 +1,493 @@
...
+    from lightning.pytorch.utilities.types import (
+        OptimizerLRSchedulerConfig,
+        LRSchedulerConfigType,
+    )
+    from lightning.pytorch.core.optimizer import LightningOptimizer
+    from lightning.fabric.utilities.types import LRScheduler
+
+    HAS_PL = True
+except ImportError:
+    HAS_PL = False
```
[the remark this answers is not in the document]
Comment (svnv-svsv-jm) [reply torchrl/lightning/loops.py:20]: But if I don't import it on the top, I can't inherit from `pl.LightningModule`, can I?

**Comment by vmoens** [comment]
@svnv-svsv-jm I think that just moving it to `trainers` without change of API would already be a good thing!

**Comment by svnv-svsv-jm** [comment]
> @svnv-svsv-jm I think that just moving it to `trainers` without change of API would already be a good thing!

I moved it by allowing `from torchrl.trainers import RLTrainingLoop`, even though the actual source code is still under `torchrl.lightning`.

**Comment by vmoens** [comment]
Not sure that cuts it unfortunately. The long term goal is that there is no `torchrl.lightning` so if we merge this as-if deprecating it will be be hard on the user side

**Comment by svnv-svsv-jm** [comment]
@vmoens No problem! Just wanted to make sure I understood correctly. So indeed I will just move `torch.lightning` to `torchrl.trainers` entirely.

**Comment by gtnv** [comment]
hi @svnv-svsv-jm, do you still plan to work on this? if not, can we close it?
